pytransform3d/trajectories.py

```diff
@@ -436,6 +509,201 @@ def pqs_from_dual_quaternions(dqs):
     return out
 
 
+
+def norm_axis_angles(a):
+    """Normalize axis-angle representation.
+
+    Parameters
+    ----------
+    a : array-like, shape (...,4)
+        Axis of rotation and rotation angle: (x, y, z, angle)
+
+    Returns
+    -------
+    a : array, shape (...,4)
+        Axis of rotation and rotation angle: (x, y, z, angle). The length
+        of the axis vector is 1 and the angle is in [0, pi). No rotation
+        is represented by [1, 0, 0, 0].
+    """
+    angle = a[:, 3]  # Extract the angle component (N,)
+    norm = np.linalg.norm(a[:, :3], axis=1)  # Compute the norm of the first three elements (axis)
+
+
+    res = np.ones_like(a)
+
+    # Create masks for elements where angle or norm is zero
+    zero_mask = (angle == 0.0) | (norm == 0.0)
+
+
+    non_zero_mask = ~zero_mask
+    res[non_zero_mask, :3] = a[non_zero_mask, :3] / norm[non_zero_mask, None]  
+
+
+    angle_normalized = norm_angle(angle)  
+
+    negative_angle_mask = angle_normalized < 0.0
+    res[negative_angle_mask, :3] *= -1.0  
+    angle_normalized[negative_angle_mask] *= -1.0  
+
+    res[non_zero_mask, 3] = angle_normalized[non_zero_mask]
+    return res
+
+
+def axis_angles_from_quaternions(qs):
+    """Compute axis-angle from quaternion.
+
+    This operation is called logarithmic map.
+
+    We usually assume active rotations.
+
+    Parameters
+    ----------
+    q : array-like, shape (...,4)
+        Unit quaternion to represent rotation: (w, x, y, z)
+
+    Returns
+    -------
+    a : array, shape (...,4)
+        Axis of rotation and rotation angle: (x, y, z, angle). The angle is
+        constrained to [0, pi) so that the mapping is unique.
+    """
+    ps = qs[:, 1:]  # Extract the vector part of the quaternion
+    p_norm = np.linalg.norm(ps, axis=1)  # Compute the norm of the vector part
+
+    # Create a mask for quaternions where p_norm is small
+    small_p_norm_mask = p_norm < np.finfo(float).eps
+
+    # Initialize the output with default values for small p_norm cases
+    result = np.zeros_like(qs)
+    result[small_p_norm_mask] = np.array([1.0, 0.0, 0.0, 0.0])
+
+    # For non-zero norms, calculate axis, clamped w, and angle
+    non_zero_mask = ~small_p_norm_mask
+    axis = np.zeros_like(ps)
+    axis[non_zero_mask] = ps[non_zero_mask] / p_norm[non_zero_mask, None]
+
+    w_clamped = np.clip(qs[:, 0], -1.0, 1.0)  # Clamp the scalar part of the quaternion
+    angle = 2.0 * np.arccos(w_clamped)
+
+    # Stack the axis and the angle together and normalize the axis-angle representation
+    result[non_zero_mask] = norm_axis_angles(np.hstack((axis[non_zero_mask], angle[non_zero_mask, None])))
+
+    return result
+
+
+def screw_parameters_from_dual_quaternions(dqs):
```

Comment:
Since this function is non-trivial when it comes to mathematical operations, would you please add at least a unittest so that a next developer wants to optimize/change/accelerate could check her work?